### ISSUE TITLE
Input widgets: add the input's type as a class or data attribute to the input's HTML tags

### ISSUE THREAD

**Issue by daattali** [rstudio/shiny#1286]
I think it would be very useful to have the input's type explicitly written out in an input's HTML.

For example, when I want to hide/disable an input, some inputs require different code (because they use third party libraries), so I need to know what the type is.  Currently I just look for things in the input's HTML that give me hints to tell me what input it is, and so far is works 100% of the time, but it feels wrong and hacky.

**Comment by jcheng5** [comment]
Adding the type name may help but you still have to grovel around in the implementation of the widget. I think we need to add hide/disable to the InputBinding API.

**Comment by daattali** [comment]
Adding those to the API is a separate issue, those were just examples of common cases when I wanted that to be available. I think if it was a standard that the .shiny-input-container element always had another class describing its input type, it could be useful

**Comment by alandipert** [comment]
I agree with Joe on this one. Whether or not it's easy for you to determine an input's type, you are still relying on implementation detail to do the stuff you want *after* you've figured out the type.

I don't consider Joe's suggestion as orthogonal; I see it as a different way to address the ultimate goal, which is supplying hide/disable functionality to end-users. The question is, who should supply it &mdash; shinyjs or shiny?

These are the two possible approaches, in my estimation:

1. Specify/publicize every level of Shiny implementation detail, starting with input type, required for you to implement shinyjs hide/disable. shinyjs becomes "right/not hacky" when it relies only on public features.
1. Add hide/disable to InputBinding API as Joe suggests. This makes the highest-level desired behavior public, and obviates that functionality in shinyjs.

The problem with going down the road of **1** is that the more internal detail of Shiny is public, the tighter the coupling between it and dependent code becomes. The tighter the coupling, the riskier changing Shiny becomes. shinyjs never becomes "right/not hacky".

So, considering we'll never be able to make shinyjs completely not-a-hack from our end, the status quo seems fine. You've taken it on yourself to continue to make sure shinyjs works (and it's a really handy package!), which is something you'd need to continue to do even if we made more internals (like the input type) public.

I made an issue to to track improving the InputBinding API, #2409, but I'm not closing this one so that we can discuss it further here if you wish.

**Comment by daattali** [comment]
I fully agree with your assessment.

My original request still stands though; I do think it would be useful to expose the type of input from the input's container's DOM. It doesn't have to be something official, doesn't need to be a public API that's documented, it would just make certain tasks easier if the container told you what input you're looking at. Over the years I've come across different StackOverflow posts where people wanted to perform some sort of action based on an input's type, and the answer has always been that you need to write special code to test for each type, and that is what I wanted to resolve with this issue.
